File: frontend_flutter/lib/shared/models/briefing_model.dart
import 'package:equatable/equatable.dart';

class BriefingModel extends Equatable {
  const BriefingModel({
    required this.leadId,
    required this.completudePct,
    this.destino,
    this.dataIda,
    this.dataVolta,
    this.numPessoas,
    this.perfil,
    this.tipoViagem,
    this.preferencias,
    this.orcamentoFaixa,
    this.passaporteValido,
    this.experienciaInternacional,
    this.resumoConversa,
  });

  final String leadId;
  final int completudePct;
  final String? destino;
  final DateTime? dataIda;
  final DateTime? dataVolta;
  final int? numPessoas;
  final String? perfil;
  final String? tipoViagem;
  final String? preferencias;
  final String? orcamentoFaixa;
  final bool? passaporteValido;
  final bool? experienciaInternacional;
  final String? resumoConversa;

  factory BriefingModel.fromJson(Map<String, dynamic> json) => BriefingModel(
        leadId: json['lead_id'] as String,
        completudePct: json['completude_pct'] as int,
        destino: json['destino'] as String?,
        dataIda: json['data_ida'] != null
            ? DateTime.parse(json['data_ida'] as String)
            : null,
        dataVolta: json['data_volta'] != null
            ? DateTime.parse(json['data_volta'] as String)
            : null,
        numPessoas: json['num_pessoas'] as int?,
        perfil: json['perfil'] as String?,
        tipoViagem: json['tipo_viagem'] as String?,
        preferencias: json['preferencias'] as String?,
        orcamentoFaixa: json['orcamento_faixa'] as String?,
        passaporteValido: json['passaporte_valido'] as bool?,
        experienciaInternacional: json['experiencia_internacional'] as bool?,
        resumoConversa: json['resumo_conversa'] as String?,
      );

  @override
  List<Object?> get props => [leadId, completudePct];
}

File: backend/app/infrastructure/persistence/models/briefing_model.py
"""
Briefing ORM Table — Infrastructure/Persistence Layer
=====================================================
SQLAlchemy model for the 'briefings' table (spec.md §4.2).
One-to-one with Lead (unique FK).

Constraints:
  - completude_pct: DB CHECK 0–100
  - qtd_pessoas: DB CHECK >= 1
  - duracao_dias: DB CHECK >= 1
  - PostgreSQL native ENUM for perfil and orcamento fields
"""
import uuid
from datetime import date
from typing import TYPE_CHECKING, Optional

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Date,
    ForeignKey,
    Integer,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import ARRAY, ENUM as SAEnum, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.domain.entities.enums import OrcamentoPerfil, PerfilViagem
from app.infrastructure.persistence.database import Base

if TYPE_CHECKING:
    from app.infrastructure.persistence.models.lead_model import LeadModel


perfil_viagem_enum = SAEnum(
    *[e.value for e in PerfilViagem],
    name="perfil_viagem_enum",
    create_type=True,
)
orcamento_perfil_enum = SAEnum(
    *[e.value for e in OrcamentoPerfil],
    name="orcamento_perfil_enum",
    create_type=True,
)

BRIEFING_REQUIRED_FIELDS = [
    "destino", "data_ida", "data_volta", "qtd_pessoas", "perfil",
    "tipo_viagem", "preferencias", "orcamento", "tem_passaporte",
]


class BriefingModel(Base):
    """
    ORM representation of a travel briefing (spec.md §4.2).
    Linked 1-to-1 with Lead. Populated incrementally by the AI agent.
    """

    __tablename__ = "briefings"

    __table_args__ = (
        CheckConstraint(
            "completude_pct BETWEEN 0 AND 100",
            name="ck_briefings_completude_range",
        ),
        CheckConstraint("qtd_pessoas IS NULL OR qtd_pessoas >= 1", name="ck_briefings_qtd_pessoas_min"),
        CheckConstraint("duracao_dias IS NULL OR duracao_dias >= 1", name="ck_briefings_duracao_min"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    lead_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("leads.id", ondelete="CASCADE"),
        unique=True,
        nullable=False,
        index=True,
    )
    destino: Mapped[Optional[str]] = mapped_column(String(255))
    origem: Mapped[Optional[str]] = mapped_column(String(255))
    data_ida: Mapped[Optional[date]] = mapped_column(Date)
    data_volta: Mapped[Optional[date]] = mapped_column(Date)
    duracao_dias: Mapped[Optional[int]] = mapped_column(Integer)
    qtd_pessoas: Mapped[Optional[int]] = mapped_column(Integer)
    perfil: Mapped[Optional[str]] = mapped_column(perfil_viagem_enum)
    tipo_viagem: Mapped[Optional[list[str]]] = mapped_column(ARRAY(String))
    preferencias: Mapped[Optional[list[str]]] = mapped_column(ARRAY(String))
    orcamento: Mapped[Optional[str]] = mapped_column(orcamento_perfil_enum)
    tem_passaporte: Mapped[Optional[bool]] = mapped_column(Boolean)
    observacoes: Mapped[Optional[str]] = mapped_column(Text)
    completude_pct: Mapped[int] = mapped_column(Integer, default=0, nullable=False)

    lead: Mapped["LeadModel"] = relationship("LeadModel", back_populates="briefing")


def calculate_completude(briefing_data: dict) -> int:
    """Calculate briefing completeness percentage (0–100)."""
    filled = sum(
        1 for field in BRIEFING_REQUIRED_FIELDS
        if briefing_data.get(field) not in (None, [], "", 0)
    )
    return round((filled / len(BRIEFING_REQUIRED_FIELDS)) * 100)
